### PR TITLE
Core/DB: Implement NULL values instead of 0 when no data is intended

### PR DESCRIPTION
**Pull request by StormBytePP**
This PR, will convert all auth and characters database 0 usage to NULL, when no data was intended.
This will enable further FKEY aditions while optimizing database (at least semantically, no longer 0 for no guid present)

### CODE REVIEW

File: src/server/game/Mails/Mail.cpp [hunk 1]
```diff
@@ -211,8 +211,8 @@ void MailDraft::SendMailTo(SQLTransaction& trans, MailReceiver const& receiver,
     stmt->setUInt16(++index, GetMailTemplateId());
     stmt->setUInt32(++index, sender.GetSenderId());
     stmt->setUInt32(++index, receiver.GetPlayerGUIDLow());
-    stmt->setString(++index, GetSubject());
```
Comment (DDuarte) [inline src/server/game/Mails/Mail.cpp]: I think these 2 strings should be kept non null. There's no reason for empty strings to be nulls

**Comment by StormBytePP** [comment]
Implemented @DDuarte change requests (reverting mail to original state) and squashed again

**Comment by StormBytePP** [comment]
There is only 1 change pending that I have no clue how to do it.
Update the UPDATES table, so this fix is not reapplied when a new install is made.

**Comment by Naios** [comment]
Just add the update entry of the update to the base dump too, see 58f4bd834dfb4dc1fc93d for details (name, hash & state = RELEASED is important, speed = 0 and the timestamp is not important).
If the entry for you file is missing in the `updates` table create one with an empty hash and start your worldserver which will calculate and insert the checksum for you without reapplying your update. Then take the entry over into your fulldump.

**Comment by StormBytePP** [comment]
Ok, I added the row to updates table, so with this, I think this is ready to be merged.

File: src/server/game/Tickets/TicketMgr.h [hunk 2]
```diff
@@ -118,7 +118,7 @@ class GmTicket
         else if (_escalatedStatus == TICKET_UNASSIGNED)
             _escalatedStatus = TICKET_ASSIGNED;
     }
-    void SetClosedBy(ObjectGuid value) { _closedBy = value; }
+    void SetClosedBy(ObjectGuid value) { _closedBy = value; _closed=true; }
```
Comment (Naios) [inline src/server/game/Tickets/TicketMgr.h]: Please insert a space between:

````c++
_closed = true
~~~~~~~^~^
```

File: src/server/database/Database/PreparedStatement.cpp [hunk 3]
```diff
@@ -82,112 +82,169 @@ void PreparedStatement::BindParameters()
...
-        statement_data.resize(index+1);
-
-    statement_data[index].data.boolean = value;
-    statement_data[index].type = TYPE_BOOL;
+    if (setNullIfZero && !value)
+        setNull(index);
+    else
+    {
+        if (index >= statement_data.size())
+            statement_data.resize(index+1);
```
Comment (Naios) [inline src/server/database/Database/PreparedStatement.cpp]: missing linebreak after `statement_data.resize(index+1);` and missing whitespace between `index+1`
-> also for all following and previous changes in this PR.

**Comment by Naios** [comment]
Seems legit, could rebase this pr and correct the two style flaws please?

**Comment by MitchesD** [comment]
summoning @StormBytePP

**Comment by jackpoz** [comment]
"This branch has conflicts that must be resolved"

**Comment by jackpoz** [comment]
I was wondering if C++ parts of this PR like the new optional argument in setType() can be merged, with no impact on the current behavior, so the sql parts can be implemented later.

**Comment by StormBytePP** [comment]
I'm currently under a very high workload so I can't attend this atm

**Comment by jackpoz** [comment]
This PR will most likely not be merged anytime soon due to the huge impact it has. It is probably better to implement base C++ part first and then migrate SQL columns in batches, maybe 1 table at time, to spot issues easily.
Please ask here to reopen this PR once you'll have time again to work on this

**Pull request closed without merging**